topi/include/topi/nn/pooling.h

```diff
@@ -84,189 +91,170 @@ inline Tensor pool_nchw(const Tensor& x,
-    return tvm::compute(
-     { batch, out_height, out_width, channel },
-      [&](Var n, Var h, Var w, Var c) {
-        return tvm::max(temp(n, h * stride_height + dheight, w * stride_width + dwidth, c),
-        { dheight, dwidth });
-      }, "tensor", "pool_max");
-  } else if (pool_type == kAvgPool) {
-    auto temp = pad(x, pad_before, pad_after, 0, "pad_temp");
-
-    auto tsum = tvm::compute(
-     { batch, out_height, out_width, channel },
-      [&](Var n, Var h, Var w, Var c) {
-        return tvm::sum(temp(n, h * stride_height + dheight, w * stride_width + dwidth, c),
-        { dheight, dwidth });
-      }, "tensor", "pool_avg");
-
-    return tvm::compute(
-     { batch, out_height, out_width, channel },
-     [&](Var n, Var h, Var w, Var c) {
-       return tsum(n, h, w, c) / (kernel_height * kernel_width);
-      }, "tensor", kElementWise);
-  } else {
-    LOG(ERROR) << "Unrecognized pool_type: " << pool_type;
-    return x;
+inline bool find_height_width(const std::string& layout,
+                              int* height_idx,
+                              int* width_idx) {
+  *height_idx = -1, *width_idx = -1;
+  int curr_idx = 0;
+  for (size_t i = 0; i < layout.size(); ++i) {
+    if ((layout[i] >= 'A' && layout[i] <= 'Z') ||
+        (layout[i] >= 'a' && layout[i] <= 'z')) {
+      if (layout[i] == 'H') {
+        if (*height_idx != -1) return false;
+        *height_idx = curr_idx;
+      } else if (layout[i] == 'W') {
+        if (*width_idx != -1) return false;
+        *width_idx = curr_idx;
+      } else if (layout[i] == 'h' || layout[i] == 'w') {
+        // do not support split on height or width, e.g., NCHW16w
+        return false;
+      }
+      ++curr_idx;
+    }
   }
+  if (*height_idx == -1 || *width_idx == -1) return false;
+  return true;
 }
 
 /*!
-* \brief Perform pooling on data
-*
-* \param x The input tensor in NCHW or NHWC order
+* \brief Perform pooling on height and width dimension of data.
+*        It decides the height and width dimension according to the layout string,
+*        in which 'W' and 'H' means width and height respectively.
+*        Width and height dimension cannot be split.
+*        For example, NCHW, NCHW16c, etc. are valid for pool,
+*        while NCHW16w, NCHW16h are not.
+*        See \a layout for more information of the layout string convention.
+* \param x The input tensor.
 * \param kernel_size Vector of two ints: {kernel_height, kernel_width}
 * \param stride_size Vector of two ints: {stride_height, stride_width}
 * \param padding_size Vector of two ints: {padding_height, padding_width}
 * \param pool_type The type of pooling operator
 * \param ceil_mode Whether to use ceil when calculating the output size
-* \param layout The input layout
-*
-* \return The output tensor in NCHW order
+* \param layout The input layout. Pooling supports any layout as long as 'H' and 'W' appear.
+*        The layout is supposed to be composed of upper cases, lower cases and numbers,
```

Comment:
In this implementation, we use "NCHWXc" to represent a 5-D layout. Would it be more clean and general to just use "NCHWc", since when implementing pooling, the exact splitting number 'X' is not needed?

Reply:
As I mentioned, when I implement similar op such as upsampling, "NCHWc" is sufficient to deal with 5-D layouts.

Reply:
I think the layout is still needed in "pooling-like" operator, but only to hint which axis are "H" and "W". This should be able to cover all cases which doesn't split "H" or "W" axis.

Reply:
Or we can remove layout and add h_axis and w_axis. But this si different from most DL framework implementation for pooling. Maybe at this point we can keep layout, and make it more general by just checking the index of "H" and "W".

Reply:
Yes I got your point. The implement here does check `H` and `W` only. For the document of the function, my concern is if we claim it is valid for something like `NCHWc`, the convention will be different from the upcoming nnvm changes. That can be confusing.

Reply:
Do we need layout convention "NCHWXc" in nnvm? This should be able to be covered by "NCHWc"?

Reply:
yes as I explained. I don't want to increase the complexity for now. as for layout transform op, it must know what `X` is.

Reply:
Just to clarify, one can transform from `NCHWc` to `NCHW` given the source shape. But will never be able to transform from `NCHW` to `NCHWc`.

Reply:
The issue here is in tvm we have no context of nnvm. Developer may be confused why we need an extra "x" number in layout, while it is not used in tvm. We can add some explanations, or add layout transform as an example, to state how developer can add extra info into layout and use it.

Reply:
all right, how about now.

Reply:
"Pooling supports any layout as long as 'H' and 'W' appear.
The layout is supposed to be composed of upper cases, lower cases and numbers, ..."
These two statements look like not quite fit with each other. Maybe we can change them to "Pooling supports any layout as long as 'H' and 'W' appear.
For example, we can add channel split factor into layout, such as NCHW16c, if this factor number is required in graph level computation."

Reply:
we don't need to mention graph level computation here, right? how about 'The layout is supposed to be composed of upper cases, lower cases and (optional) numbers' ?

Reply:
Yes. This looks fine.

Reply:
here we go